p2p/service/gossip_listener.go

```diff
@@ -78,8 +78,41 @@ func (l *Listener) Stop() {
 	l.wg.Wait()
 }
 
-func (l *Listener) listenToGossip(ctx context.Context, dataHandler GossipDataHandler, gossipChannel chan GossipMessage, stop chan struct{}) {
-	l.WithContext(ctx).Info("start listening")
+func (l *Listener) listenToGossip(ctx context.Context, dataHandler GossipDataHandler, gossipChannel chan GossipMessage, stop chan struct{}, channel string) {
+	l.WithContext(ctx).With().Info("start listening to gossip", log.String("protocol", channel))
+
+	// fill the channel with tokens to limit number of concurrent routines
+	tokenChan := make(chan struct{}, l.config.MaxGossipRoutines)
+	for i := 0; i < l.config.MaxGossipRoutines; i++ {
+		tokenChan <- struct{}{}
+	}
+
+	handleMsg := func(ctx context.Context, data GossipMessage) {
+		// get a token to create a new channel
+		l.WithContext(ctx).With().Info("waiting for available slot for gossip handler",
+			log.Int("available_slots", len(tokenChan)),
+			log.Int("total_slots", cap(tokenChan)))
+		if len(tokenChan) == 0 {
+			l.WithContext(ctx).Error("no available slots for gossip handler, blocking")
+		}
+		<-tokenChan
+
+		l.WithContext(ctx).With().Info("got gossip message, forwarding to data handler",
+			log.String("protocol", channel),
+			log.Int("queue_length", len(gossipChannel)))
+		if !l.syncer.ListenToGossip() {
+			// not accepting data
+			l.WithContext(ctx).Info("not currently listening to gossip, dropping message")
+			return
+		}
+		go func() {
```

Comment:
this design will cause many gouroutines to be created and waiting on `tokenChan`. I wonder if this is the desired behaviour? another option is to block listenToGossip until we have a worker ready

Reply:
agree, I updated this along the lines you suggested